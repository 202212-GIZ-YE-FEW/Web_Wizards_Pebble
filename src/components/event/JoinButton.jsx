import { useState } from "react";

import { useAuthContext } from "@/context/AuthContext";
import useFirestore from "@/firebase/firestore";

const JoinButton = ({ eventId }) => {
    const { user } = useAuthContext();
    const [isJoining, setIsJoining] = useState(false);
    const [isLeaving, setIsLeaving] = useState(false);
    const [isJoined, setIsJoined] = useState(false);

    const [error, setError] = useState();
    const eventsHook = useFirestore("events");

    const handleJoinClick = async () => {
        try {
            setIsJoining(true);
            const event = await eventsHook.getDocumentById(eventId);
            const attendees = event.attendees || [];

            if (!attendees.includes(user.displayName)) {
                const updatedAttendees = [...attendees, user.displayName];
                await eventsHook.updateDocumentById(eventId, {
                    attendees: updatedAttendees,
                });
            }

            setIsJoining(false);
        } catch (error) {
            console.error("Error joining Eevent", error);
            setError(error);
            setIsJoining(false);
        }
    };

    const handleLeaveClick = async () => {
        try {
            setIsLeaving(true);
            const event = await eventsHook.getDocumentById(eventId);
            const attendees = event.attendees || [];

            const updatedAttendees = attendees.filter(
                (attendee) => attendee !== user.displayName
            );
            await eventsHook.updateDocumentById(eventId, {
                attendees: updatedAttendees,
            });
            setIsJoined(false);

            setIsLeaving(false);
        } catch (error) {
            console.error("Error leaving event", error);
            setError(error);
            setIsLeaving(false);
        }
    };

    return (
        <>
            <div className='py-12 order-3 lg:order-none px-4'>
                <button
                    onClick={handleJoinClick}
                    disabled={isJoining}
                    className='w-full bg-primary-200 focus:outline-none text-white font-semibold rounded-lg text-md px-32 py-3 '
                >
                    Join
                </button>
            </div>

            <div className='py-12 order-3 lg:order-none px-4'>
                <button
                    onClick={handleLeaveClick}
                    disabled={isJoining}
                    className='w-full bg-primary-200 focus:outline-none text-white font-semibold rounded-lg text-md px-32 py-3 '
                >
                    Leave
                </button>
            </div>
        </>
    );
};

export default JoinButton;
