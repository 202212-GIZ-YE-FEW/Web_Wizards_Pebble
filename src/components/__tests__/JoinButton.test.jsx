import renderer from "react-test-renderer";

import JoinButton from "../event/JoinButton";

it("renders correctly", () => {
    const tree = renderer.create(<JoinButton />).toJSON();
    expect(tree).toMatchSnapshot();
});
