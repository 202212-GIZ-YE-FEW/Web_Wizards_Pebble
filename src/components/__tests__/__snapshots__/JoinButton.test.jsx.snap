// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="py-12 order-3 lg:order-none px-4"
>
  <button
    className="w-full bg-primary-200 focus:outline-none text-white font-semibold rounded-lg text-md px-32 py-3 "
  >
    Join
  </button>
</div>
`;
